materialize-slack/main.go

```diff
@@ -0,0 +1,300 @@
+	}
+
+	api, apiErr := cfg.buildAPI()
+	if apiErr != nil {
+		return nil, apiErr
+	}
+
+	for _, binding := range req.Materialization.Bindings {
+		var res resource
+		if err := pf.UnmarshalStrict(binding.ResourceConfigJson, &res); err != nil {
+			return nil, fmt.Errorf("parsing resource config: %w", err)
+		}
+
+		var channelInfo, err = api.ConversationInfo(res.Channel)
+		if err != nil {
+			return nil, fmt.Errorf("error getting channel: %s, %w", res.Channel, err)
+		}
+
+		if !channelInfo.IsMember {
+			if err := api.JoinChannel(res.Channel); err != nil {
+				return nil, fmt.Errorf("error joining channel: %s, %w", res.Channel, err)
+			}
+		}
+	}
+	return &pm.Response_Applied{ActionDescription: "materialize-slack does not modify channels"}, nil
+}
+
+// Transactions implements the DriverServer interface.
+func (driver) NewTransactor(ctx context.Context, open pm.Request_Open) (pm.Transactor, *pm.Response_Opened, error) {
+
+	var cfg config
+	if err := pf.UnmarshalStrict(open.Materialization.ConfigJson, &cfg); err != nil {
+		return nil, nil, fmt.Errorf("parsing endpoint config: %w", err)
+	}
+
+	api, err := cfg.buildAPI()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var bindings []*binding
+	for _, b := range open.Materialization.Bindings {
+		var res resource
+		if err := json.Unmarshal(b.ResourceConfigJson, &res); err != nil {
+			return nil, nil, fmt.Errorf("unable to parse resource config: %w", err)
+		}
+		bindings = append(bindings, &binding{
+			resource:   res,
+			collection: string(b.Collection.Name),
+			spec:       b,
+		})
+	}
+
+	var transactor = &transactor{
+		api:      api,
+		bindings: bindings,
+	}
+
+	return transactor, &pm.Response_Opened{}, nil
+}
+
+type transactor struct {
+	api      *SlackAPI
+	bindings []*binding
+}
+
+type binding struct {
+	spec       *pf.MaterializationSpec_Binding
+	resource   resource
+	collection string
+}
+
+func buildDocument(b *binding, keys, values tuple.Tuple) map[string]interface{} {
+	var document = make(map[string]interface{})
+
+	// Add the keys to the document.
+	for i, value := range keys {
+		var propName = b.spec.FieldSelection.Keys[i]
+		document[propName] = value
+	}
+
+	// Add the non-keys to the document.
+	for i, value := range values {
+		var propName = b.spec.FieldSelection.Values[i]
+
+		if raw, ok := value.([]byte); ok {
+			document[propName] = json.RawMessage(raw)
+		} else {
+			document[propName] = value
+		}
+	}
+	return document
+}
+
+func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	for it.Next() {
+		panic("Load should never be called for materialize-slack")
+	}
+	return nil
+}
+
+func (t *transactor) Store(it *pm.StoreIterator) (pm.StartCommitFunc, error) {
+
+	for it.Next() {
+		var b = t.bindings[it.Binding]
+		var parsed = buildDocument(b, it.Key, it.Values)
+		var tsStr, tsOk = parsed["ts"].(string)
+		var text, textOk = parsed["text"].(string)
+		var blocks = parsed["blocks"].(json.RawMessage)
```

Comment:
I _think_ this is ok if `blocks` isn't present in the output from `buildDocument`? `nil` can be coerved into a `json.RawMessage` (which is a []byte)...and later `string([]byte(nil))` I assume is `""`? I'm not 100% sure on this so if you haven't tried it, it might not be a bad idea to test it out.

Reply:
narrator: it was not okay. fixed it